snapcraft/internal/lifecycle/_clean.py

```diff
@@ -36,56 +36,51 @@ def _reverse_dependency_tree(config, part_name):
     return dependents
 
 
-def _clean_part_and_all_dependents(part_name, step, config, staged_state,
-                                   primed_state):
-    # Obtain the reverse dependency tree for this part. Make sure all
-    # dependents are cleaned.
-    dependents = _reverse_dependency_tree(config, part_name)
-    dependent_parts = {p for p in config.all_parts
-                       if p.name in dependents}
-    for dependent_part in dependent_parts:
-        dependent_part.clean(staged_state, primed_state, step)
+def _clean_part(part_name, step, config, staged_state, primed_state):
+    if step == common.COMMAND_ORDER[-1]:
+        template = 'Cleaning {step} step for {part}'
+    else:
+        template = 'Cleaning {step} step (and all subsequent steps) for {part}'
 
-    # Finally, clean the part in question
+    logger.info(template.format(step=step, part=part_name))
     config.parts.clean_part(part_name, staged_state, primed_state, step)
 
+    # If we just cleaned the root of a dependency tree, we need to mark all
+    # its dependents as dirty so they require cleaning
+    return mark_dependents_dirty(part_name, step, config)
 
-def _verify_dependents_will_be_cleaned(part_name, clean_part_names, step,
-                                       config):
-    # Get the name of the parts that depend upon this one
-    dependents = config.parts.get_dependents(part_name)
-    additional_dependents = []
 
-    # Verify that they're either already clean, or that they will be cleaned.
-    if not dependents.issubset(clean_part_names):
-        for part in config.all_parts:
-            if part.name in dependents and not part.is_clean(step):
-                humanized_parts = formatting_utils.humanize_list(
-                    dependents, 'and')
-                additional_dependents.append(part_name)
+def mark_dependents_dirty(part_name, cleaned_step, config):
+    dependents = _reverse_dependency_tree(config, part_name)
+    dependent_parts = {p for p in config.all_parts
```

Comment:
which makes this confusing given that we just got the reverse dependency tree for `part_name` and now we need to iterate again to get the _real_ dependents?

Reply:
Yeah, that only gives us part names, and in order to clean we need the actual `PluginHandler` instances.